### PR TITLE
Support exchange spooling on Azure Blob Storage

### PR DESCRIPTION
**Pull request by linzebing**
## Description

> Is this change a fix, improvement, new feature, refactoring, or other?

New feature

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

trino-exchange-plugin

> How would you describe this change to a non-technical end user or system administrator?

This enables users to set up exchange spooling on Azure Blob Storage.

## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

(x) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: plugin/trino-delta-lake/pom.xml [hunk 1]
```diff
@@ -288,7 +288,7 @@
         <dependency>
             <groupId>com.azure</groupId>
             <artifactId>azure-core</artifactId>
-            <version>1.13.0</version>
+            <version>1.25.0</version>
```
Comment (losipiuk) [inline plugin/trino-delta-lake/pom.xml:291]: nit: version bump in separate preparatory commit?

File: plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/ExchangeAzureConfig.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import io.airlift.units.MaxDataSize;
+import io.airlift.units.MinDataSize;
+
+import javax.validation.constraints.NotNull;
+
+import static io.airlift.units.DataSize.Unit.MEGABYTE;
+
+public class ExchangeAzureConfig
+{
+    private String azureStorageConnectionString = System.getenv("AZURE_STORAGE_CONNECTION_STRING");
```
Comment (losipiuk) [inline plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/ExchangeAzureConfig.java:29]: nit: make it `Optional<String>`

Comment (losipiuk) [reply plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/ExchangeAzureConfig.java:29]: Also I think we should not depend on system variable here. I do not think we ever do that. It is easier to reason about cluster deployment if config files are the sole source of truth when it comes to what configuration is actually used (or make it as close to that as possible).

File: plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/local/LocalFileSystemExchangeStorage.java [hunk 3]
```diff
@@ -168,6 +168,7 @@ private static List<Path> listPaths(URI directory, Predicate<Path> predicate)
         private static final int INSTANCE_SIZE = ClassLayout.parseClass(LocalExchangeStorageReader.class).instanceSize();
 
         private final Queue<ExchangeSourceFile> sourceFiles;
+
```
Comment (losipiuk) [inline plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/local/LocalFileSystemExchangeStorage.java:171]: unrelated

File: plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/s3/S3FileSystemExchangeStorage.java [hunk 4]
```diff
@@ -188,17 +188,15 @@ public ListenableFuture<Void> createEmptyFile(URI file)
     }
 
     @Override
-    public ListenableFuture<Void> deleteRecursively(URI uri)
```
Comment (losipiuk) [inline plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/s3/S3FileSystemExchangeStorage.java:191]: unrelated changes here - extract preparatory commit.

File: plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/ExchangeAzureConfig.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    private DataSize azureStorageBlockSize = DataSize.of(4, MEGABYTE);
+
+    public String getAzureStorageConnectionString()
+    {
+        return azureStorageConnectionString;
+    }
+
+    @Config("exchange.azure.storage-connection-string")
```
Comment (losipiuk) [inline plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/ExchangeAzureConfig.java:37]: nit: should we drop `storage-` prefix from both properties?

File: plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java [hunk 6]
```diff
@@ -0,0 +1,455 @@
...
+                throws IOException
+        {
+            if (closed) {
+                return null;
+            }
+            if (sliceInput != null && sliceInput.isReadable()) {
+                return sliceInput.readSlice(sliceInput.readInt());
+            }
+            ExchangeSourceFile sourceFile = sourceFiles.poll();
+            if (sourceFile != null) {
```
Comment (losipiuk) [inline plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java]:
```java
            if (sourceFile == null) {
                close();
                return null;
            }
```
and then without `else {`

File: plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java [hunk 7]
```diff
@@ -0,0 +1,455 @@
...
+
+            String blockId = Base64.getEncoder().encodeToString(UUID.randomUUID().toString().getBytes(StandardCharsets.UTF_8));
+            ListenableFuture<Void> uploadFuture = toListenableFuture(blockBlobAsyncClient.stageBlock(blockId, Flux.just(slice.toByteBuffer()), slice.length()).toFuture());
+            multiPartUploadFutures.add(uploadFuture);
+            blockIds.add(blockId);
+            return uploadFuture;
+        }
+
+        @Override
+        public ListenableFuture<Void> finish()
```
Comment (losipiuk) [inline plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java:509]: would that make sense to mark Writer as `finishing` so we throw an error if `finish` is called twice (or we can return previous future for second call).

(can be a followup - if to be done, same applies for s3 implementation)

File: plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java [hunk 8]
```diff
@@ -0,0 +1,455 @@
...
+            multiPartUploadFutures.forEach(future -> future.cancel(true));
+
+            // No explicit way to delete staged blocks; uncommitted blocks are automatically deleted after 7 days
+            return immediateVoidFuture();
+        }
+
+        @Override
+        public long getRetainedSize()
+        {
+            return INSTANCE_SIZE;
```
Comment (losipiuk) [inline plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java:564]: nit: there are two lists also - but I guess it is fine to skip those as size should be negligible.

File: plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java [hunk 9]
```diff
@@ -0,0 +1,455 @@
...
+    @Override
+    public List<FileStatus> listFiles(URI dir)
+            throws IOException
+    {
+        ImmutableList.Builder<FileStatus> builder = ImmutableList.builder();
+        try {
+            for (BlobItem blobItem : listObjects(dir)) {
+                if (blobItem.isPrefix() != Boolean.TRUE) {
+                    builder.add(new FileStatus(new URI(dir.getScheme(), dir.getUserInfo(), dir.getHost(), -1, PATH_SEPARATOR + blobItem.getName(), null, dir.getFragment()).toString(),
+                            blobItem.getProperties().getContentLength()));
```
Comment (losipiuk) [inline plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java:176]:
```java
                    builder.add(new FileStatus(
                            new URI(dir.getScheme(), dir.getUserInfo(), dir.getHost(), -1, PATH_SEPARATOR + blobItem.getName(), null, dir.getFragment()).toString(),
                            blobItem.getProperties().getContentLength()));
```

File: plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java [hunk 10]
```diff
@@ -273,79 +279,185 @@ private static boolean isDirectory(URI uri)
...
+
+                BlockBlobAsyncClient blockBlobAsyncClient = blobServiceAsyncClient
+                        .getBlobContainerAsyncClient(getContainerName(currentFile.getFileUri()))
+                        .getBlobAsyncClient(getPath(currentFile.getFileUri()))
+                        .getBlockBlobAsyncClient();
+                Optional<SecretKey> secretKey = currentFile.getSecretKey();
+                if (secretKey.isPresent()) {
+                    blockBlobAsyncClient = blockBlobAsyncClient.getCustomerProvidedKeyAsyncClient(new CustomerProvidedKey(secretKey.get().getEncoded()));
+                }
+                for (int i = 0; i < readableBlocks && fileOffset < fileSize; ++i) {
```
Comment (losipiuk) [inline plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java:424]: Do we need `&& fileOffset < fileSize` here? Isn't how `readableBlocks` is computed enough to not go over file size?

File: plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java [hunk 11]
```diff
@@ -273,79 +279,185 @@ private static boolean isDirectory(URI uri)
...
-                String blobName = getPath(sourceFile.getFileUri());
-                Optional<SecretKey> secretKey = sourceFile.getSecretKey();
-                BlobClient blobClient = blobServiceClient.getBlobContainerClient(containerName).getBlobClient(blobName);
-                if (secretKey.isPresent()) {
-                    blobClient = blobClient.getCustomerProvidedKeyClient(new CustomerProvidedKey(secretKey.get().getEncoded()));
+
+            if (sliceSize < 0) {
+                sliceSize = sliceInput.readInt();
+            }
+            Slice data = sliceInput.readSlice(sliceSize);
```
Comment (losipiuk) [inline plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java:333]: It is not obvious that `sliceInput` will have at least `sliceSize` bytes at this point (after single call to `fillBuffer`) worth a comment.

**Review by losipiuk — COMMENTED** [review]
LGTM - some nitpicks.

File: plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java (hunk 6 above)
Comment (linzebing) [reply plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java]: Replaced with parallel reader

File: plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java (hunk 7 above)
Comment (linzebing) [reply plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java:509]: Actually in `ExchangeSink`'s protocol, for finish() `This method is guaranteed not be called more than once.`.

File: plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java (hunk 8 above)
Comment (linzebing) [reply plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java:564]: Good catch. Added accounting for `blockIds`, but not sure how to get estimated size of a `ListenableFuture<Void>`

File: plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java (hunk 10 above)
Comment (linzebing) [reply plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java:424]: We need. readableBlocks is calculated based on the remaining space in the buffer, but it's possible that the remaining file content is even shorter

File: plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java (hunk 11 above)
Comment (linzebing) [reply plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java:333]: I agree, but I find it a little awkward to add comment here as the code seems to explain itself better.

File: plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java [hunk 12]
```diff
@@ -0,0 +1,570 @@
...
+            blockBlobAsyncClient = blockBlobAsyncClient.getCustomerProvidedKeyAsyncClient(new CustomerProvidedKey(secretKey.get().getEncoded()));
+        }
+        return new AzureExchangeStorageWriter(blockBlobAsyncClient, blockSize);
+    }
+
+    @Override
+    public boolean exists(URI file)
+            throws IOException
+    {
+        return blobServiceClient.getBlobContainerClient(getContainerName(file)).getBlobClient(getPath(file)).exists();
```
Comment (arhimondr) [inline plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java:145]: Is `blobServiceClient.getBlobContainerClient` cheap? Do we need to cache it?

File: plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java [hunk 13]
```diff
@@ -0,0 +1,570 @@
...
+    }
+
+    private PagedFlux<BlobItem> listObjectsRecursively(URI dir)
+    {
+        checkArgument(isDirectory(dir), "listObjectsRecursively called on file uri %s", dir);
+
+        String containerName = getContainerName(dir);
+        String directoryPath = getPath(dir);
+
+        return blobServiceAsyncClient.getBlobContainerAsyncClient(containerName).listBlobsByHierarchy(null, (new ListBlobsOptions()).setPrefix(directoryPath));
```
Comment (arhimondr) [inline plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java:240]: How many objects does it return in a single batch by default?

File: plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java [hunk 14]
```diff
@@ -0,0 +1,570 @@
...
+        @Override
+        public boolean isFinished()
+        {
+            return closed;
+        }
+
+        @Override
+        public void close()
+        {
+            if (closed) {
```
Comment (arhimondr) [inline plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java:374]: Does the `close` method have to be synchronized?

File: plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java [hunk 15]
```diff
@@ -0,0 +1,570 @@
...
+                }
+                for (int i = 0; i < readableBlocks && fileOffset < fileSize; ++i) {
+                    int length = (int) min(blockSize, fileSize - fileOffset);
+
+                    int finalBufferFill = bufferFill;
+                    FluentFuture<Void> downloadFuture = FluentFuture.from(toListenableFuture(blockBlobAsyncClient.downloadWithResponse(new BlobRange(fileOffset, (long) length), null, null, false).toFuture()))
+                            .transformAsync(response -> toListenableFuture(response.getValue().collectList().toFuture()), directExecutor())
+                            .transform(byteBuffers -> {
+                                int offset = finalBufferFill;
+                                for (ByteBuffer byteBuffer : byteBuffers) {
```
Comment (arhimondr) [inline plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java:435]: Is it possible to have the client directly write bytes into a destination buffer (to avoid double copying?)

File: plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java (hunk 14 above)
Comment (linzebing) [reply plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java:374]: Nice catch, I forgot

File: plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java (hunk 13 above)
Comment (linzebing) [reply plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java:240]: Default is 5000.

File: plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java (hunk 12 above)
Comment (linzebing) [reply plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java:145]: It doesn't have any network ops. Just class construction, negligible cost?

Comment (linzebing) [reply plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java:145]: good advice

File: plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java (hunk 15 above)
Comment (linzebing) [reply plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java:435]: I don't see an easy way, as Azure SDK doesn't provide any kind of callback/transformer interface to do so.

File: plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java (hunk 12 above)
Comment (arhimondr) [reply plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java:145]: If it's just a matter of creating a wrapper object than it's fine

**Review by arhimondr — APPROVED** [review]

**Comment by linzebing** [comment]
CI: https://github.com/trinodb/trino/pull/12298

**Comment by linzebing** [comment]
CI: https://github.com/trinodb/trino/issues/12300

**Review by losipiuk — APPROVED** [review]
